### ISSUE TITLE
rekeying child, when pfs=yes, silently ignores DH in esp=... and uses DH from IKE exchange

### ISSUE THREAD

**Issue by cagney** [libreswan/libreswan#1265]
given:
```
conn east
	pfs=yes
	ike=aes_gcm;modp2048
	esp=aes_gcm;modp3072
conn west
	pfs=yes
	ike=aes_gcm;modp2048
	esp=aes_gcm;modp4096
```
the modp is silently but deliberately stripped:
```
| ESP/AH string values: AES_GCM_16-NONE-MODP4096
| constructing ESP/AH proposals with all DH removed  for west (loading config)
| converting proposal AES_GCM_16-NONE-MODP4096 to ikev2 ...
| forcing IKEv2 ESP AES_GCM_16 ENCRYPT transform low-to-high key lengths: 128 256
| ...  ikev2_proposal: 1:ESP=AES_GCM_C_128+AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
"west": Child SA proposals (connection add):
"west":   1:ESP=AES_GCM_C_128+AES_GCM_C_256-NONE-NONE-ENABLED+DISABLED
```
at the time of the rekey, the IKE SAs DH group is added:
```
"west" #3: initiating rekey to replace Child SA #2
"west" #3: sent CREATE_CHILD_SA request to rekey IPsec SA
...
"east" #3: proposal 1:ESP=AES_GCM_C_128-MODP2048-ENABLED SPI=ad45777a chosen from remote proposals 1:ESP:ENCR=AES_GCM_C_128;DH=MODP2048;ESN=ENABLED[first-match]
...
"west" #3: initiator rekeyed Child SA #2 using #1; IPsec tunnel [192.0.1.0-192.0.1.255:0-65535 0] -> [192.0.2.0-192.0.2.255:0-65535 0] {ESP/ESN=>0xb4ef456a <0xad45777a xfrm=AES_GCM_16_128-NONE-MODP2048 DPD=passive}
```
minimally, there should be a warning about this happening (when pfs=no, that happens).

ref #1253

**Comment by letoams** [comment]
On Mon, 11 Sep 2023, cagney wrote:

> Subject: [libreswan/libreswan] rekeying child, when pfs=yes,
>     silently ignores DH in esp=... and uses DH from IKE exchange (Issue #1265)

> the modp is silently but deliberately stripped:

I am more and more leaning towards just refusing to load such settings.
But someone will need to do some stupid interop somewhere I guess :/

Maybe at least throw a warning saying NOT RECOMMENDED.

**Comment by cagney** [comment]
> I am more and more leaning towards just refusing to load such settings.

Yes. and it turns out we're doing that; just silently.

My problem is that I'm trying to reproduce either #1253 or #1259 and they need things messed up.

**Comment by cagney** [comment]
Prior to 9f6f83556c089f4e2c8b979631445aff770168be the rekey emitted all the proposals from the ESP= line and when PFS missing DH was filled in from IKE.

**Comment by cagney** [comment]
I'm marking this as a regression.  Without a way to force proposing nonsense there isn't really an easy way to work-around #1253

**Comment by cagney** [comment]
perhaps a combination like:
```
msdh-downgrade=yes
pfs=no
```
should propose no-pfs and any pfs algorithms specified

**Comment by cagney** [comment]
pfs-peer-workaround= or pfs=dh* + impair to allow wierd responder

**Comment by cagney** [comment]
Old behaviour:

`pfs=no`
- verbosely strip an DH from `esp=` line
- since `esp=` proposal has no DH, emit proposal without DH

`pfs=yes`
- leave DH in the `esp=` line
- if `esp=` proposal has no DH, emit proposal with IKE DH
- if `esp=` proposal has DH, emit proposal with ESP's DH
- if `msdh-downgrade=yes` also emit proposal without DH

**Comment by cagney** [comment]
New behaviour:

`pfs=no`
- verbosely strip DH from `esp=` line
- since `esp=` proposal has no DH, emit proposal without DH

`pfs=yes`
- leave DH in `esp=` line
- regardless of DH in `esp=` proposal, emit proposal with IKE DH
- if `msdh-downgrade=yes` also emit proposal with no DH

**Comment by cagney** [comment]
So `pfs=yes` needs to change:

- require either all `esp=` or no `esp=` proposals have DH; this esp=aes-sha1,aes-sha1-dh16 is invalid

to:

- warn (reject?) that any DH in `esp=` line will be replaced by DH from IKE

**Comment by letoams** [comment]
On Wed, 18 Oct 2023, cagney wrote:

> New behaviour:
> 
> pfs=no
>
>  *  verbosely strip DH from esp= line
>  *  since esp= proposal has no DH, emit proposal without DH
> 
> pfs=yes
>
>  *  leave DH in esp= line
>  *  regardless of DH in esp= proposal, emit proposal with IKE DH

While there is a reason to do this for IKEv2, there really isn't for
IKEv1? Is IKEv1 behaviour changed as well?

>  *  if msdh-downgrade=yes also emit proposal with no DH

**Comment by cagney** [comment]
> While there is a reason to do this for IKEv2, there really isn't for IKEv1? Is IKEv1 behaviour changed as well?

This is all IKEv2

**Comment by bleve** [comment]
So my suggestion is to go back to pre v4.6 behaviour and drop ms-dh-downgrade option if pluto is build without DH2 support (which is default).

**Comment by cagney** [comment]
from #1337 there needs to be ways to specify:

- IKE AUTH specifying ENCR+INTEG (ignoring DH)

- new child specifying ENCR+INTEG+IKE DH
- new child specifying ENCR+INTEG+custom DH
- new child specifying ENCR+INTEG+no-DH

- rekey existing ESP ENCR+INTEG+previous DH
- rekey existing ESP ENCR+INTEG+IKE DH
- rekey existing ESP ENCR+INTEG+custom DH
- rekey specifying custom ENCR+INTEG+DH

**Comment by cagney** [comment]
Trying

pfs=no: add algs, or OLD
pfs=yes: add algs(AUTH), algs+IKE(NEW), or OLD+IKE(REKEY)
pfs-peer-workaround=yes: add algs[-dh](NEW, REKEY), 
msdh=yes: append none to proposals

with the weird edge case pfs=no;workaround=yes is interpreted as only add workaround

| exch    | esp | pfs | msdh | workaround | proposal | notes |
|-|-|-|-|-|-|-|
|||||||
|  AUTH  |algs | no | no     |  no   | algs                 | 
|  AUTH  |algs | no | no     |  yes   | algs                 | 
|  AUTH  |algs | no | yes     |  no   | algs                 | 
|  AUTH  |algs | no | yes     |  yes   | algs                 | 
|  AUTH  |algs | yes | no      |  no | algs                | 
|  AUTH  |algs | yes | no  |  yes | algs                |
|  AUTH  |algs | yes | yes     |  no   | algs                 | 
|  AUTH  |algs | yes | yes     |  yes   | algs                 |  
|  AUTH  |algs-dh | no | no     |  no   | invalid                 | 
|  AUTH  |algs-dh  | no | no     |  yes   | algs?         |   v4.5      |
|  AUTH  |algs-dh  | no | yes     |  no   | invalid                 | 
|  AUTH  |algs-dh  | no | yes     |  yes   |  algs  |    v4.5     |
|  AUTH  |algs-dh  | yes | no      |  no | algs                |  
|  AUTH  |algs-dh  | yes | no  |  yes | algs                |    
|  AUTH  |algs-dh  | yes | yes     |  no   | algs                 | 
|  AUTH  |algs-dh  | yes | yes     |  yes   | algs                 |
|||||||
|  NEW | algs  | no | no    |  no    | algs-none  |
|  NEW | algs  | no | no    |  yes    | algs-none  |  
|  NEW | algs | no | yes     |  no   | algs-none |  
|  NEW | algs | no | yes     |  yes   | algs-none |  
|  NEW | algs | yes | no     |  no  | algs-IKE | 
|  NEW | algs | yes | no     |  yes  | algs-IKE,algs-none | 
|  NEW | algs | yes | yes     |  no  | algs-IKE+none    | 
|  NEW | algs | yes | yes     |  yes  | algs-IKE+none,algs-none    | silly | 
|  NEW  | algs-dh | no | no     |  no   | invalid  |  
|  NEW  | algs-dh | no | no     |  yes   | algs-dh |v4.5 |
|  NEW  | algs-dh | no | yes     |  no   | invalid  |  |
|  NEW  | algs-dh | no | yes     |  yes   | algs-hd+none | v4.5 |
|  NEW  | algs-dh | yes | no     |  no   | algs-IKE  |  
|  NEW  | algs-dh | yes | no     |  yes   | algs-IKE,algs-dh  |
|  NEW    | algs-dh | yes | yes    |  no    | algs-IKE+none   |
|  NEW    | algs-dh | yes | yes    |  yes    | algs-IKE+none,algs-dh+none | 
|||||||
|  REKEY | algs  | no | no    |  no    | OLD-none  |
|  REKEY | algs  | no | no    |  yes    | OLD-none,algs-none  |  
|  REKEY | algs | no | yes     |  no   | OLD-none |   
|  REKEY | algs | no | yes     |  yes   | OLD-none,algs-none   |
|  REKEY | algs | yes | no     |  no  | OLD-IKE|
|  REKEY | algs | yes | no     |  yes  | OLD-IKE,algs-none |
|  REKEY | algs | yes | yes     |  no  | OLD-IKE+none,algs-none   |
|  REKEY | algs | yes | yes     |  yes  | OLD-IKE+none,algs-none   |
|  REKEY  | algs-dh | no | no     |  no   | invalid  |
|  REKEY  | algs-dh | no | no     |  yes   | algs-dh| v4.5|
|  REKEY  | algs-dh | no | yes     |  no   | invalid  | 
|  REKEY  | algs-dh | no | yes     |  yes   | algs-dh+non | v4.5 |
|  REKEY  | algs-dh | yes | no     |  no   | OLD-IKE  |  
|  REKEY  | algs-dh | yes | no     |  yes   | OLD-IKE,algs-dh  | 
|  REKEY    | algs-dh | yes | yes    |  no    | OLD-IKE+none   |  
|  REKEY    | algs-dh | yes | yes    |  yes    | OLD-IKE+none,algs-dh+none |

**Comment by cagney** [comment]
I think the scenario to focus on is rekeying a Child SA created uding IKE_AUTH with:
```
ike=IKE_ALG...-IKE_DH...
esp=ESP_ALG...-ESP_DH...
```
where ike establishes IKE_ALG-IKE_DH and the IKE_AUTH child establishes ESP_ALG, and when the Diffie-Hellman groups for IKE and ESP (IKE_DH... and ESP_DH...) do not intersect.

| |IKE_AUTH<br>propose| IKE_AUTH<br>accept | REKEY<br>propose|
|-|-|-|-|
|v4.5|ESP_ALG... | ESP_ALG | ESP_ALG...-ESP_DH... |
|v4.6|ESP_ALG..|ESP_ALG|ESP_ALG-IKE_DH|

since IKE_DH is not in ESP_DH... (and vis vis), they can't interop

**Comment by cagney** [comment]
by default:

- child always gets IKE DH whe PFS
- warn and ignore esp=

when pfs-rekey-workaround=yes

- child gets esp= dh

need to update table to see if it works

**Comment by cagney** [comment]
Table updated, two questions:

- should `ms-dh-downgrade=` and `pfs-rekey-workaround=` be mutually exclusive; I'm going to assume yes so the two are not interdependent
- given esp=ALGS-DH should the second child propose ALGS-IKE instead; would break with v4.x behaviour

**Comment by cagney** [comment]
To answer:

> given esp=ALGS-DH should the second child propose ALGS-IKE instead; would break both v4.5- and v4.6+ behaviour

Consider two connections with `pfs=yes` that can share their IKE SA, one with `esp=ALGS-DH` and the other with `esp=ALGS`.  With v4.6+ the order that the two connections establish affects what is negotiated.  While the `esp=ALGS` connection always ends up with the IKE's DH, the `esp=ALGS-DH`, under v4.6+ ends up with either:

- the IKE's DH when `algs=ALGS-DH` is negotiated during IKE AUTH and then rekeyed
- the DH from `algs=ALGS-DH` when negotiated using CREATE_CHILD_SA (i.e., second child)

which, I think, really shouldn't happen (under v4.5- `algs=ALGS-DH` always ended up with DH)

Hence I'd suggest calling the option something like `child-uses-ike-dh=` and default to yes, or almost `pfs=ike`

**Comment by bleve** [comment]
I find pfs=yes|no|ike approach better. 

Did you mean: pfs=yes is work-arounds and pfs=ike is forced pfs with ike-dh?

**Comment by cagney** [comment]
> Did you mean: pfs=yes is work-arounds and pfs=ike is forced pfs with ike-dh?

yes, but I've this nagging feeling that it might be interpretred as use keying material from the IKE exchange i.e., a variation on pfs=no

**Comment by bleve** [comment]
I don't think so. People don't understand protocols and this is just a matter of documentation.

**Comment by cagney** [comment]
So something like:

- pfs=no, use keying material derived from the IKE exchange
- pfs=ike, negotiate and generate new keying material using the IKE SA's DH algorithm
- pfs=yes, negotiate and generate new keying material using a DH algorithm specified as part of esp= (or the IKE SA's DH algorithm when unspecified)

the next problem is that existing examples have pfs=yes so the odds are that will be specified which isn't the desire; sigh

**Comment by letoams** [comment]
On Thu, 23 Nov 2023, Tuomo Soini wrote:

> I find pfs=yes|no|ike approach better.

I would rather not do this. It looks too "legitimate". I really want
the option set with like "workaround" in the name to show that whatever
they are trying is unwise, and not normal.

**Comment by antonyantony** [comment]
On Thu, Nov 23, 2023 at 08:02:13AM -0800, cagney wrote:
> So something like:
> 
> - pfs=no, use keying material derived from the IKE exchange
> - pfs=ike, negotiate and generate new keying material using the IKE SA's DH algorithm

are there any other IPsec implementations that support pfs=ike and next one  
where it inherit from IKE? Initially it sounded logical to me, now oveer 
time, I wonder if we pinoeer in this?

**Comment by cagney** [comment]
> I wonder if we pinoeer in this?

very much so, starting in v4.6

**Comment by cagney** [comment]
Time to cut our losses and revert the behaviour back to 4.5.
We've no factual basis for the claim that using the IKE DH is recommended and, hence, the end users config should be ignored.

**Comment by antonyantony** [comment]
On Fri, Nov 24, 2023 at 08:13:17AM -0800, The Libreswan Vault wrote:
> Closed #1265 as completed via 18f40759e1e38da8ba6b86564f7733f4c93bb69a.

Thanks for digging into this.

I invite to add the test case names in the issues when it is relevant. I can 
imagine test case would get renamed... However, if there are test cases 
mentiond, I would be more motivated to look into it or poke around with the 
test case.

Do Pluto, as a rekey initiator, try with second DH? when the there multiple 
DH possible? Either from IKE or specified in the Child config?

When pluto is the responder what error would it send DH mismatches?

At last IETF there were cases mentioned that make sense not to use DH for 
PFP short lived SAs.

**Comment by cagney** [comment]
> I invite to add the test case names in the issues when it is relevant. I can imagine test case would get renamed... However, if there are test cases mentiond, I would be more motivated to look into it or poke around with the test case.

The pfs-* tests play around with the edges, but nothing yet explicit.

> Do Pluto, as a rekey initiator, try with second DH? when the there multiple DH possible? Either from IKE or specified in the Child config?

No.  See #1260 and #1259.  5.0 fixed INVALID_KE (sent back and logged) response to the rekey request, but still doesn't then make use of it.  Workaround is then to adjust the `esp=` line manually.  Look for invalid-ke tests.

IKE rekeys only allow the original DH.
